### PR TITLE
Improve `tests/subtests`

### LINKED ISSUE #159: Reinstate sub-tests
On the advice and recommendation of @ConorMacBride the `tests.test_subtests` have been temporarily disabled in #150.

I've created this issue as a reminder to reinstate the sub-tests.

This issue can also serve as a focus for dev discussion on how we best go about doing this.

For reference:

- See this [comment](https://github.com/matplotlib/pytest-mpl/pull/150#issuecomment-1132565957)
- PR #134 
- PR #145

### PR DESCRIPTION
**Pull request by ConorMacBride**
Now the `tests/subtests` can be updated/regenerated automatically without needing to manually adjust the baseline data.

To create the baseline images and hashes run:
```sh
MPL_UPDATE_BASELINE=1 tox -e mpl35
```

Then, to create the baseline summaries run:
```sh
MPL_UPDATE_SUMMARY=1 tox -e mpl35
```

Finally check the `git diff` and run the standard tests before committing:
```sh
tox -e mpl35
```

I will add documentation for this in a README file in `tests/subtests`.

All changes are isolated to within `tests/subtests` except I've added `Pillow` (imported as `PIL`) as a dependency. (Matplotlib already requires this package and #150 will need it also.) And I have made some minor changes to `tox.ini`.

Interestingly, removing the matplotlib version number from the PNGs means that **all** the hashes are identical across matplotlib 3.3, 3.4 and 3.5!! (I guess this isn't too surprising since the plots are very basic.)

- [x] Add documentation

Closes #159

### CODE REVIEW

File: tests/subtests/subtest.py [hunk 1]
```diff
@@ -15,86 +15,99 @@ def plot(line, **kwargs):
 
 # hash match
 
-@pytest.mark.mpl_image_compare()
+@pytest.mark.image
+@pytest.mark.hash
+@pytest.mark.mpl_image_compare(savefig_kwargs={'metadata': {'Software': None}})
```
Comment (bjlittle) [inline tests/subtests/subtest.py:20]: @ConorMacBride @Cadair  Should I also adopt this pattern as a default in #160 ?

Seems like a sensible thing to do, right?

Comment (ConorMacBride) [reply tests/subtests/subtest.py:20]: Yeah, I think that would be a good idea. Then we can remove the `metadata` setting from this file.

File: tests/subtests/test_subtest.py [hunk 2]
```diff
@@ -23,14 +26,28 @@
...
+UPDATE_SUMMARY = os.getenv("MPL_UPDATE_SUMMARY") is not None  # baseline summaries
+
+# When updating baseline summaries, replace parts of status_msg with regex.
+# See helpers.apply_regex for more information.
+REGEX_PATHS = [
+    str(Path(__file__).parent),  # replace all references to baseline files
+    os.path.realpath(tempfile.gettempdir()),  # replace all references to output files
+]
+REGEX_STRS = [
+    r'RMS Value: ',
```
Comment (ConorMacBride) [inline tests/subtests/test_subtest.py:49]: @bjlittle, in #150 this will need changed to `r'RMS: '`.

File: tests/subtests/README.rst [hunk 3]
```diff
@@ -0,0 +1,78 @@
+Testing ``pytest-mpl`` using the ``tests/subtests``
+**************************************************
+
+``pytest-mpl`` can output JSON summaries (``--mpl-generate-summary=json``) which contain lots of machine readable information relating to the internal state of the plugin while it was run.
+This test module (``test_subtest.py``) runs the test file ``subtest.py`` multiple times with different combinations of z
```
Comment (bjlittle) [inline tests/subtests/README.rst:5]:
```suggestion
This test module (``test_subtest.py``) runs the test file ``subtest.py`` multiple times with different combinations of 
```

File: tests/subtests/README.rst [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+:``hmatch``: Hash comparison must pass, so same hash in baseline and result hash libraries.
+
+:``hdiff``: Hash comparison must fail, so baseline hash is set to the same as the result hash except the first four characters are changed to ``d1ff``.
+
+:``hmissing``: Baseline hash must be missing, so baseline hash is deleted from the baseline hash library but not the result hash library.
+
+**Image comparison status flags:**
+
+:``imatch``: Image comparison must pass, so correct image in included in the baseline directory.
```
Comment (bjlittle) [inline tests/subtests/README.rst:35]:
```suggestion
:``imatch``: Image comparison must pass, so correct image is included in the baseline directory.
```

**Review by bjlittle — APPROVED** [review]
@ConorMacBride Lovely. Looking forward to using this straight away for #150 👍

**Comment by ConorMacBride** [comment]
pre-commit.ci autofix

**Pull request merged**